pkg/cloud/services/ec2/service.go

```diff
@@ -32,6 +32,8 @@ type Service struct {
 
 	// SSMClient is used to look up the official EKS AMI ID
 	SSMClient ssmiface.SSMAPI
+	// isMock is used to return current mock service instead of creating new one, for testing
+	isMock bool
```

Comment:
You don't need this. Could you take a look at other tests which create this service? :) You can see that we are mocking the scope, which is then, in turn, used to set up these services.

Reply:
hey @Skarlso 
thanks for comment.  I saw servicefactory pattern in controller like 
```go
// https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/exp/controllers/awsmachinepool_controller.go#L74
func (r *AWSMachinePoolReconciler) getEC2Service(scope scope.EC2Scope) services.EC2Interface {
	if r.ec2ServiceFactory != nil {
		return r.ec2ServiceFactory(scope)
	}

	return ec2.NewService(scope)
}
```
the issue here is that EC2Service has to create new EC2Service inside the `ReconcileLaunchTemplate` method, so I can't apply serviceFactory to it. Can we reuse EC2Service instead of creating new one?   Any suggestion? 🙏 
```golang
// https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/ec2/launchtemplate.go#L64
func (s *Service) ReconcileLaunchTemplate(
	scope scope.LaunchTemplateScope,
	canUpdateLaunchTemplate func() (bool, error),
	runPostLaunchTemplateUpdateOperation func() error,
) error {
        ...
	ec2svc := NewService(scope.GetEC2Scope())
        ...
}


type Service struct {
	scope     scope.EC2Scope
	EC2Client ec2iface.EC2API

	// SSMClient is used to look up the official EKS AMI ID
	SSMClient ssmiface.SSMAPI
}

// NewService returns a new service given the ec2 api client.
func NewService(clusterScope scope.EC2Scope) *Service {
	return &Service{
		scope:     clusterScope,
		EC2Client: scope.NewEC2Client(clusterScope, clusterScope, clusterScope, clusterScope.InfraCluster()),
		SSMClient: scope.NewSSMClient(clusterScope, clusterScope, clusterScope, clusterScope.InfraCluster()),
	}
}
```

Reply:
another idea is to create factory in EC2Service, but we still need new fields.
```go
func (s *Service) new(clusterScope scope.EC2Scope) *Service {
	var ec2Client ec2iface.EC2API
	if s.ec2ClientFactory != nil {
		ec2Client = s.ec2ClientFactory()
	} else {
		ec2Client = scope.NewEC2Client(clusterScope, clusterScope, clusterScope, clusterScope.InfraCluster())
	}
	var ssmClient ssmiface.SSMAPI
	if s.ssmClientFactory != nil {
		ssmClient = s.ssmClientFactory()
	} else {
		ssmClient = scope.NewSSMClient(clusterScope, clusterScope, clusterScope, clusterScope.InfraCluster())
	}

	return &Service{
		scope:     clusterScope,
		EC2Client: ec2Client,
		SSMClient: ssmClient,
	}
}
```

Reply:
Okay, so you don't need any of that because you have the scope and the ec2Mock EXPECTs. And the scope is mocked and you can control the ec2 service responses from there. From example, `GetEC2Scope` will return `InfraCluster` from the MachinePool scope.

So when you set things up here:

```go
		InfraCluster:       mcps,
```

you should be able to return something canned. I can take a look at that later if you don't find anything. :)

Reply:
Ah crap, I see what you mean. And I see now that we do that with the Service Factory pattern:

```go
			ec2ServiceFactory: func(scope.EC2Scope) services.EC2Interface {
				return ec2Svc
			},
```

Okay. Would you mind please switching this isMock to that then please? Cheers!

Reply:
Thanks, and sorry it took so long to reply. Eventually, I'll rework this into using proper injection pattern. But that's not gonna happen in this PR. :) Cheers.

Reply:
hi @Skarlso ,

sorry for late reply,   I refactored it in [refactor to service factory pattern](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4388/commits/fc49a1ee80b2c8e70402d06d0b60767c151f0622). If you have time, please take a look.

if you have already done rework in another PR,  I would be happy to follow that. 😄